### PR TITLE
[WIP] fix --check-untyped-defs for MyPy

### PR DESCRIPTION
**Pull request by simonjayhawkins**
**tl;dr No need for review!**

no intention of merging this so using patterns that probably won't be accepted.

PR is for visibility and reference only.

using http://blog.zulip.org/2016/10/13/static-types-in-python-oh-mypy/ as a suitable template for a project plan. resolution of #27568 would come before too much work on adding type hints in general.

There is quite a bit of work involved to enable check-untyped-defs globally so need to keep the phases described in "Fully annotating a large codebase" section as out-of-scope for now.

I do not expect to complete this in the near term, so am using Python 3.6 type annotations for variables (better formatting with black and no need to initialize variables to None) and mypy 0.720 for checking (less false positives)

One of the advantages of enabling check-untyped-defs globally, is the additional checking of PRs.

However, for as long as I can keep this branch in sync with master, this additional checking is effectively being performed.

The goal is to cover as much ground in the shortest possible time. The piecemeal approach of submitting separate PRs was, IMO, not allowing us to see the wood from the trees. The extent of the work involved (and changes required) won't be known till the majority of modules have been covered.

So far, nothing major has surfaced. If latent bugs are discovered could always break bits off and submit for review along the way.

### CODE REVIEW

File: pandas/_libs/indexing.pyi [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# flake8: noqa
```
Comment (jbrockmendel) [inline pandas/_libs/indexing.pyi]: im not sure if it helps, but indexing.pyx is one of the few cython files that _can_ be made into valid python

File: pandas/io/excel/_util.py [hunk 2]
```diff
@@ -85,7 +86,7 @@ def _excel2num(x):
     return index - 1
 
 
-def _range2cols(areas):
+def _range2cols(areas: str) -> List[int]:
```
Comment (jbrockmendel) [inline pandas/io/excel/_util.py]: are you going to make separate PRs for stuff like this that works immediately?

File: pandas/_libs/indexing.pyi (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/_libs/indexing.pyi]: Cool. Thanks.

my original plan, using http://blog.zulip.org/2016/10/13/static-types-in-python-oh-mypy/ as a suitable template for a project plan was to include the 'skeleton' stubs in the preliminary phase.

the reasoning being that if the cython files were python files, mypy would have the same information that is being made available with the skeleton stubs (i.e. without any type hints).

This was also the reason not to include the types in the stubs at this stage, to keep the preliminary phase separate from phase 1.

However, adding the stubs is increasing the mypy errors reported by check-untyped-defs, so was making no progress trying to get check-untyped-defs enabled globally.

so in order to keep things manageable, the preliminary phase needs to be split in two. I've left the stubs in this branch for now but do not intend to add any more to this branch.

File: pandas/io/excel/_util.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/io/excel/_util.py]: not sure of the approach at this time. not really made much impact on the complete task yet.

a couple of things to consider:

to realize the intended benefit of additional PR checking, we'll need to enable the check-untyped-defs globally and create a blacklist.
 
In order to remove modules from the blacklist quicker, i've started with the modules containing the least errors, but these would be the best choices for community engagement, so happy if that's done by the community and just use this branch as a reference for a potential solution.

File: pandas/_libs/indexing.pyi (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/indexing.pyi]: Makes sense, thanks.  Also FWIW, cython understands the py36 style variable annotations even in py35, see e.g. lib.memory_usage_of_objects

File: pandas/core/computation/engines.py [hunk 3]
```diff
@@ -120,12 +114,8 @@ def _evaluate(self):
...
             return ne.evaluate(s, local_dict=scope)
         except KeyError as e:
-            # python 3 compat kludge
-            try:
-                msg = e.message
-            except AttributeError:
-                msg = str(e)
-            raise UndefinedVariableError(msg)
+            name = str(e)
+            raise UndefinedVariableError(name=name)
```
Comment (simonjayhawkins) [inline pandas/core/computation/engines.py]: cc @jbrockmendel 

if this path was tested (and reachable) it would have raised `__init__() missing 1 required positional argument: 'is_local'`

fix here is to make `is_local` optional keyword.

Comment (jbrockmendel) [reply pandas/core/computation/engines.py]: Thanks for checking this out.  Does this merit a separate PR?

Comment (simonjayhawkins) [reply pandas/core/computation/engines.py]: > Does this merit a separate PR?

will do. 

on master mypy --check-untyped-defs (with mypy 0.720) reports 372 errors. 304 of these still need to be investigated!

so not yet in a position to be able to prioritize fixes.

Comment (jbrockmendel) [reply pandas/core/computation/engines.py]: Also, there was another recent case where there was an un-tested exception we were calling incorrectly.  Will typing prevent that?

Comment (simonjayhawkins) [reply pandas/core/computation/engines.py]: with check-untyped-defs set (and the relevant module not blacklisted) then in theory, yes.

otherwise, if any type annotations are added to the calling function/method then the body of the function will be checked, so should also report errors.

so check-untyped-defs can serve two purposes, checking functions while still in the process of add type annotations and also checking new functions added without type annotations.

the second scenario would be unnecessary if we used disallow-untyped-defs. That's along way off though.

**Comment by simonjayhawkins** [comment]
@jbrockmendel I would appreciate another pair of eyes on https://github.com/pandas-dev/pandas/pull/28339/commits/a398d83f96c4b3c92006308c51b6bedaebe5e91b

**Comment by simonjayhawkins** [comment]
_get_counts is already typed 
```
def _get_counts(
    values_shape: Tuple[int],
    mask: Optional[np.ndarray],
    axis: Optional[int],
    dtype=float,
) -> Union[int, np.ndarray]:
```

**Comment by jbrockmendel** [comment]
> I would appreciate another pair of eyes on a398d83

The edits in the tests look good.

The annotation for nanmean looks right.  I _think_ `values` should be `Union[np.ndarray, ExtensionArray]`.

_FuncType and _F make things marginally _less_ clear to me, but I trust that you know what you're doing on that front.

**Comment by simonjayhawkins** [comment]
nanmean is decorated by disallow and bottleneck_switch, so this is needed to preserve the types. see https://mypy.readthedocs.io/en/latest/generics.html#declaring-decorators

**Comment by simonjayhawkins** [comment]
> The annotation for nanmean looks right. I _think_ `values` should be `Union[np.ndarray, ExtensionArray]`.

added.

**Comment by simonjayhawkins** [comment]
> @jbrockmendel I would appreciate another pair of eyes on [a398d83](https://github.com/pandas-dev/pandas/commit/a398d83f96c4b3c92006308c51b6bedaebe5e91b)

Thanks.

**Comment by simonjayhawkins** [comment]
Improved check_untyped_defs in mypy 0.740

from the blog http://mypy-lang.blogspot.com/2019/10/mypy-0740-released.html ...

"The self argument to methods without annotations is now correctly given the object’s type (instead of Any) when using check_untyped_defs (PR 7530).

This allows substantially better checking of untyped methods as well as enabling inference of attribute types in some situations where they would not have been determined in the past.

Additionally, we now suppress most “needs type annotation” errors caused by assigning empty collections (i.e, data = []), in untyped functions."

this "substantially better checking" increases the number of errors reported significantly. so the number of outstanding errors to resolve is now...

<details>

    121 pandas\tseries\offsets.py
     67 pandas\core\indexes\base.py
     58 pandas\plotting\_matplotlib\core.py
     51 pandas\core\base.py
     49 pandas\core\generic.py
     44 pandas\core\arrays\base.py
     39 pandas\io\pytables.py
     34 pandas\io\parsers.py
     29 pandas\core\arrays\datetimelike.py
     25 pandas\io\stata.py
     22 pandas\core\internals\blocks.py
     19 pandas\util\testing.py
     19 pandas\core\indexes\datetimelike.py
     18 pandas\core\internals\managers.py
     17 pandas\core\indexes\category.py
     17 pandas\core\computation\pytables.py
     16 pandas\core\indexes\interval.py
     13 pandas\io\json\_json.py
     13 pandas\core\apply.py
     12 pandas\core\computation\ops.py
     11 pandas\io\sas\sas_xport.py
     11 pandas\core\internals\construction.py
     11 pandas\core\groupby\generic.py
     10 pandas\io\excel\_openpyxl.py
      8 pandas\core\indexes\range.py
      8 pandas\core\groupby\base.py
      8 pandas\_version.py
      7 pandas\core\indexes\period.py
      6 pandas\core\resample.py
      6 pandas\core\indexes\datetimes.py
      6 pandas\core\groupby\groupby.py
      6 pandas\core\arrays\sparse\array.py
      5 pandas\core\arrays\interval.py
      5 pandas\_config\config.py
      4 pandas\io\packers.py
      4 pandas\io\excel\_xlsxwriter.py
      4 pandas\core\window\ewm.py
      3 pandas\plotting\_matplotlib\misc.py
      3 pandas\io\html.py
      3 pandas\io\excel\_base.py
      3 pandas\core\indexes\multi.py
      3 pandas\core\groupby\grouper.py
      3 pandas\compat\chainmap.py
      2 pandas\core\indexing.py
      2 pandas\core\indexes\timedeltas.py
      2 pandas\core\indexes\numeric.py
      2 pandas\core\computation\scope.py
      2 pandas\core\computation\expr.py
      2 pandas\core\algorithms.py
      1 pandas\plotting\_misc.py
      1 pandas\io\sas\sas7bdat.py
      1 pandas\io\formats\style.py
      1 pandas\io\formats\csvs.py
      1 pandas\io\excel\_xlwt.py
      1 pandas\io\common.py
      1 pandas\core\window\common.py
      1 pandas\core\reshape\merge.py
      1 pandas\core\groupby\ops.py
      1 pandas\core\frame.py
      1 pandas\core\arrays\string_.py
      1 pandas\core\arrays\period.py
      1 pandas\core\arrays\numpy_.py

</details>

File: pandas/io/common.py [hunk 4]
```diff
@@ -550,9 +550,10 @@ def __init__(
         super().__init__(file, mode, zipfile.ZIP_DEFLATED, **kwargs)
 
     def write(self, data):
+        # TODO: fixme, self.filename could be None
+        # error: Argument 1 to "writestr" of "ZipFile" has incompatible type
+        #  "Optional[str]"; expected "Union[str, ZipInfo]"  [arg-type]
         archive_name = self.filename
```
Comment (simonjayhawkins) [inline pandas/io/common.py]: @WillAyd I _think_ there could be a latent bug here, but masked by #19827

File: pandas/io/excel/_base.py [hunk 5]
```diff
@@ -771,15 +772,14 @@ def _value_with_fmt(self, val):
 
         return val, fmt
 
-    @classmethod
-    def check_extension(cls, ext):
+    def check_extension(self, ext):
```
Comment (simonjayhawkins) [inline pandas/io/excel/_base.py]: @WillAyd 

AFAICT the implementation works but mypy is reporting `pandas\io\excel\_base.py:781:54: error: "Callable[[ExcelWriter], Tuple[str, ...]]" has no attribute "__iter__" (not iterable)  [attr-defined]` (now L780)

although `supported_extensions` is a class variable, and it is defined in the abstract base class as a property purely to ensure that a subclass defines it. mypy thinks that it may also be a property in a subclass and this would be invalid code.

asked on gitter https://gitter.im/python/typing?at=5db30d223d669b28a0d62c6f but response was that mypy is correct.

`check_extension` does not need to be a class method and mypy does not report errors with a regular method.

any reason not to change from classmethod?

Comment (WillAyd) [reply pandas/io/excel/_base.py]: Well the downside is that this is technically exposed as part of the API, so moving from a class method to an instance method would be a breaking change.

For the subclasses does defining `supported_extensions` as a property not fix it? Glanced at the mypy convo that was what I picked up but might have misinterpreted

Comment (simonjayhawkins) [reply pandas/io/excel/_base.py]: > For the subclasses does defining `supported_extensions` as a property not fix it?

IIUC that is why mypy is reporting an error, since defining `supported_extensions` as a property in the subclasses leads to a runtime error...

The other option is not to define `supported_extensions` as an abstract property.

```
__________________________________________________________________________________________ test_excel_multiindex_index[.xls] __________________________________________________________________________________________ 

ext = '.xls'

    def test_excel_multiindex_index(ext):
        # MultiIndex as index works so assert no error #9794
        cols = MultiIndex.from_tuples(
            [("site", ""), ("2014", "height"), ("2014", "weight")]
        )
        df = DataFrame(np.random.randn(3, 10), index=cols)
        with ensure_clean(ext) as path:
>           df.to_excel(path, index=False)

pandas\tests\io\excel\test_xlwt.py:42:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas\core\generic.py:2296: in to_excel
    engine=engine,
pandas\io\formats\excel.py:735: in write
    writer = ExcelWriter(_stringify_path(writer), engine=engine)  # type: ignore
pandas\io\excel\_xlwt.py:25: in __init__
    super().__init__(path, mode=mode, **engine_kwargs)
pandas\io\excel\_base.py:714: in __init__
    self.check_extension(ext)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.io.excel._xlwt._XlwtWriter'>, ext = 'xls'

    @classmethod
    def check_extension(cls, ext):
        """checks that path's extension against the Writer's supported
        extensions.  If it isn't supported, raises UnsupportedFiletypeError."""
        if ext.startswith("."):
            ext = ext[1:]
>       if not any(ext in extension for extension in cls.supported_extensions):
E       TypeError: 'property' object is not iterable

pandas\io\excel\_base.py:781: TypeError
```

Comment (WillAyd) [reply pandas/io/excel/_base.py]: This code is pretty strange. Does defining `a` as an abstract `classmethod` instead of a `property` work? What we have in the base class doesn't work unless you overwrite the property with a class variable, so yea I think makes sense for mypy to flag this

Comment (simonjayhawkins) [reply pandas/io/excel/_base.py]: > What we have in the base class doesn't work unless you overwrite the property with a class variable

IIUC what we have in the base class is the construct to ensure that subclasses define `supported_extensions`, so i'm not sure what you mean when you say what is in the base class doesn't work since it is an abstract class that cannot be instantiated.

so yes "What we have in the base class doesn't work unless you overwrite the property with a class variable" but isn't that the reason for using

```
    @property
    @abc.abstractmethod
    def supported_extensions(self):
        """Extensions that writer engine supports."""
        pass
```
instead of just 
```
supported_extensions = None
```

Comment (WillAyd) [reply pandas/io/excel/_base.py]: `property` I believe works for only for instances; we are just mixing instance / class methods in a strange way and mypy is letting us know that. If you ignore the abstract thing for a minute, it certainly would be an error to call a class method that depends on an instance method

Sure you could just remove the `abstract` requirement altogether for `supported_extensions` and make it a class variable, but that would lose the abstract nature of it. You can alternately just convert the property to a class method and that keeps things consistent while passing validation

```python
from abc import ABCMeta, abstractmethod
from typing import Tuple


class A(metaclass=ABCMeta):
    @classmethod
    @abstractmethod
    def a(self) -> Tuple[int, ...]:
        pass

    @classmethod
    def b(cls) -> None:
        for x in cls.a():
            pass


class B(A):

    @classmethod
    def a(self):
        return (1, 2, 3)
```

Comment (simonjayhawkins) [reply pandas/io/excel/_base.py]: > You can alternately just convert the property to a class method and that keeps things consistent while passing validation

Thanks @WillAyd for looking into this.

i _think_ that this suggestion would also be considered a breaking change.

```
___________________________________________________________________________________ TestExcelWriterEngineTests.test_register_writer ___________________________________________________________________________________ 

self = <pandas.tests.io.excel.test_writers.TestExcelWriterEngineTests object at 0x00000202B49F78C8>

    def test_register_writer(self):
        # some awkward mocking to test out dispatch and such actually works
        called_save = []
        called_write_cells = []
    
        class DummyClass(ExcelWriter):
            called_save = False
            called_write_cells = False
            supported_extensions = ["xlsx", "xls"]
            engine = "dummy"
    
            def save(self):
                called_save.append(True)
    
            def write_cells(self, *args, **kwargs):
                called_write_cells.append(True)
    
        def check_called(func):
            func()
            assert len(called_save) >= 1
            assert len(called_write_cells) >= 1
            del called_save[:]
            del called_write_cells[:]
    
        with pd.option_context("io.excel.xlsx.writer", "dummy"):
            register_writer(DummyClass)
>           writer = ExcelWriter("something.xlsx")

pandas\tests\io\excel\test_writers.py:1257:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas\io\excel\_base.py:705: in __init__
    self.check_extension(ext)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.tests.io.excel.test_writers.TestExcelWriterEngineTests.test_register_writer.<locals>.DummyClass'>, ext = 'xlsx'

    @classmethod
    def check_extension(cls, ext: str) -> bool:
        """checks that path's extension against the Writer's supported
        extensions.  If it isn't supported, raises UnsupportedFiletypeError."""
        if ext.startswith("."):
            ext = ext[1:]
>       if not any(ext in extension for extension in cls.supported_extensions()):
E       TypeError: 'list' object is not callable

pandas\io\excel\_base.py:771: TypeError
```

Comment (simonjayhawkins) [reply pandas/io/excel/_base.py]: > Well the downside is that this is technically exposed as part of the API, so moving from a class method to an instance method would be a breaking change.

"None of the methods and properties are considered public." see https://dev.pandas.io/docs/reference/api/pandas.ExcelWriter.html?highlight=excelwriter#pandas.ExcelWriter

Comment (simonjayhawkins) [reply pandas/io/excel/_base.py]: > AFAICT the implementation works but mypy is reporting `pandas\io\excel\_base.py:781:54: error: "Callable[[ExcelWriter], Tuple[str, ...]]" has no attribute "__iter__" (not iterable) [attr-defined]` (now L780)
> 
> although `supported_extensions` is a class variable, and it is defined in the abstract base class as a property purely to ensure that a subclass defines it. mypy thinks that it may also be a property in a subclass and this would be invalid code.
> 
> asked on gitter https://gitter.im/python/typing?at=5db30d223d669b28a0d62c6f but response was that mypy is correct.

xref https://github.com/python/mypy/issues/7760

**Comment by jbrockmendel** [comment]
does dropping py35 make this actionable?

**Comment by simonjayhawkins** [comment]
I think could start breaking bits off.. the issue is the mypy version i'm using.

mypy wasn't pinned before #29431, so could not know when conda would resolve to newer version. so used latest to keep ahead of the curve.

The problem is using 0.740 this task has become virtually impossible to achieve without additional (community) support https://github.com/pandas-dev/pandas/pull/28339#issuecomment-543684846

started with around 400 errors to invesigate, couple of months to get just over half way, with 0.740 went up to 800.

**Comment by simonjayhawkins** [comment]
The first step is to enable --check-untyped-defs globally and lockdown the passing modules. @WillAyd has just given the green light https://gitter.im/pydata/pandas?at=5dc5b425ca778c1fbfc9f3cc

**Comment by jbrockmendel** [comment]
looks like this is green.  does that mean ready for review?

**Comment by jreback** [comment]
@simonjayhawkins I think would be great to push a PR with things you have typed in this one

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins I think would be great to push a PR with things you have typed in this one

first step is to get #29493 merged to lockdown passing modules.

**Comment by simonjayhawkins** [comment]
ci no longer running on this draft branch. @datapythonista any idea?

**Comment by datapythonista** [comment]
> ci no longer running on this draft branch. @datapythonista any idea?

I haven't worked with draft PRs, is CI expected to run in them?

**Comment by simonjayhawkins** [comment]
I

> I haven't worked with draft PRs, is CI expected to run in them?

it had been. I think on a couple of occasions in the past it may have not triggered, but the last 3 pushes in a row have not triggered.

The last push that triggered was 200 commits.

maybe later today if I get the chance, I'll create a new dummy PR as draft and see what happens if I don't see any other draft PRs triggering ci.

**Comment by simonjayhawkins** [comment]
@datapythonista ci working again.

**Comment by simonjayhawkins** [comment]
closing to reduce noise on notifications when keeping in sync with master

**Comment by simonjayhawkins** [comment]
reopening temporarily see https://github.com/pandas-dev/pandas/pull/33274#issuecomment-611932062

File: pandas/io/excel/_base.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/io/excel/_base.py]: I dont know how either mypy or abc.abstractmethod would handle this, but it isn't _that_ difficult to implement a class-property (i previously looked into this for Timestamp/Timedelta min/max)
